examples/Cassie/cassie_rbt_state_estimator.cc

```diff
@@ -1,140 +1,981 @@
+
+    // Get the heel spring deflection direction and magnitude
+    const Vector3d spring_rest_dir(1, 0, 0);
+    double heel_spring_angle =
+        acos(r_sol_wrt_heel_base.dot(spring_rest_dir) /
+            (r_sol_wrt_heel_base.norm() * spring_rest_dir.norm()));
+    Vector3d r_rest_dir_cross_r_hs_to_sol = spring_rest_dir.cross(
+        r_sol_wrt_heel_base);
+    int spring_deflect_sign = (r_rest_dir_cross_r_hs_to_sol(2) >= 0) ? 1 : -1;
+    if (i == 0)
+      *left_heel_spring = spring_deflect_sign * heel_spring_angle
+                         - spring_rest_offset;
+    else
+      *right_heel_spring = spring_deflect_sign * heel_spring_angle
+                          - spring_rest_offset;
+  }  // end for
+}
+
+
+void CassieRbtStateEstimator::AssignImuValueToOutputVector(
+    OutputVector<double>* output, const cassie_out_t& cassie_out) const {
+  const double* imu_d = cassie_out.pelvis.vectorNav.linearAcceleration;
+  output->SetIMUAccelerationAtIndex(0, imu_d[0]);
+  output->SetIMUAccelerationAtIndex(1, imu_d[1]);
+  output->SetIMUAccelerationAtIndex(2, imu_d[2]);
+}
+
+void CassieRbtStateEstimator::AssignActuationFeedbackToOutputVector(
+    OutputVector<double>* output, const cassie_out_t& cassie_out) const {
+  // Copy actuators
+  output->SetEffortAtIndex(actuator_index_map_.at("hip_roll_left_motor"),
+                           cassie_out.leftLeg.hipRollDrive.torque);
+  output->SetEffortAtIndex(actuator_index_map_.at("hip_yaw_left_motor"),
+                           cassie_out.leftLeg.hipYawDrive.torque);
+  output->SetEffortAtIndex(actuator_index_map_.at("hip_pitch_left_motor"),
+                           cassie_out.leftLeg.hipPitchDrive.torque);
+  output->SetEffortAtIndex(actuator_index_map_.at("knee_left_motor"),
+                           cassie_out.leftLeg.kneeDrive.torque);
+  output->SetEffortAtIndex(actuator_index_map_.at("toe_left_motor"),
+                           cassie_out.leftLeg.footDrive.torque);
+
+  output->SetEffortAtIndex(actuator_index_map_.at("hip_roll_right_motor"),
+                           cassie_out.rightLeg.hipRollDrive.torque);
+  output->SetEffortAtIndex(actuator_index_map_.at("hip_yaw_right_motor"),
+                           cassie_out.rightLeg.hipYawDrive.torque);
+  output->SetEffortAtIndex(actuator_index_map_.at("hip_pitch_right_motor"),
+                           cassie_out.rightLeg.hipPitchDrive.torque);
+  output->SetEffortAtIndex(actuator_index_map_.at("knee_right_motor"),
+                           cassie_out.rightLeg.kneeDrive.torque);
+  output->SetEffortAtIndex(actuator_index_map_.at("toe_right_motor"),
+                           cassie_out.rightLeg.footDrive.torque);
+}
+
+void CassieRbtStateEstimator::AssignNonFloatingBaseStateToOutputVector(
+    OutputVector<double>* output, const cassie_out_t& cassie_out) const {
+  // Copy the robot state excluding floating base
+  // TODO(yuming): check what cassie_out.leftLeg.footJoint.position is.
```

Comment:
Is this TODO complete or still TODO? Seems important.

Reply:
This is still a TODO. We cannot check it in simulation, because we create `cassie_out` ourselves. I'm wondering what `cassie_out.leftLeg.footJoint.position` in Cassie hardware.